test/utils/test_github_utils.py

```diff
@@ -0,0 +1,147 @@
+                                 'mode': github_utils._EXEC_MODE,
+                                 'type': 'blob',
+                                 'content': '#!/usr/bin/env python\n'}]}),
+                          _USERNAME, _PASSWORD),
+                      mock.call(
+                          requests.post, _POST_TREE_URL,
+                          json.dumps({'base_tree': 'root_sha',
+                                      'tree':
+                                      [{'path': 'main',
+                                        'mode': github_utils._TREE_MODE,
+                                        'type': 'tree',
+                                        'sha': 'fake-tree-sha'},
+                                       {'path': 'util',
+                                        'mode': github_utils._TREE_MODE,
+                                        'type': 'tree',
+                                        'sha': 'fake-tree-sha'}]}),
+                          _USERNAME, _PASSWORD),
+                      mock.call(requests.post, _POST_COMMIT_URL,
+                                json.dumps({'tree': 'fake-tree-sha',
+                                            'parents': ['fake-commit-sha'],
+                                            'message': 'fake-message'}),
+                                _USERNAME, _PASSWORD),
+                      mock.call(requests.patch, _PATCH_REF_URL,
+                                json.dumps({'force': True,
+                                            'sha': 'fake-commit-sha'}),
+                                _USERNAME, _PASSWORD)]
+
+
+def _requests_side_effect(http_method, url, req_data=None, username=None,
+                          password=None):
+    """Side effect function for mocking of get_request_content"""
+    if http_method == requests.post:
+        if (url == _POST_TREE_URL and username == _USERNAME and
+                password == _PASSWORD):
+            return {'sha': 'fake-tree-sha'}
+        if (url == _POST_COMMIT_URL and username == _USERNAME and
+                password == _PASSWORD):
+            return {'sha': 'fake-commit-sha'}
+    if http_method == requests.get:
+        if url == _GET_BRANCH_URL:
+            return _FAKE_BRANCH
+        if (len(url) >= len(_GET_TREE_URL) and
+                url[:len(_GET_TREE_URL)] == _GET_TREE_URL):
+            return _REMOTE_TREES[url[len(_GET_TREE_URL):]]
+    if not (http_method == requests.patch and url == _PATCH_REF_URL and
+            username == _USERNAME and password == _PASSWORD):
+        raise Exception('Bad HTTP request at {0}'.format(url))
+
+
+def _imm_subdirs_side_effect(curr_dir):
+    """Must mock this function so order of subdirectories is known"""
+    if curr_dir == _FAKE_LOCAL:
+        return ['main', 'util']
+    elif curr_dir == os.path.join(_FAKE_LOCAL, 'main'):
+        return ['subdir']
+    return []
+
+
+@mock.patch('pipeline.utils.github_utils._get_immediate_subdirs',
```

Comment:
The function that I'm testing has to use os functions to walk through the specified local directory, and I have to mock the function for listing directory items so that I know what order the tree will be built in. What do you mean by mocking the pipeline?